erts/emulator/beam/erl_bits.c

```diff
@@ -1729,18 +1729,31 @@ erts_bs_private_append_checked(Process* p, Eterm bin, Uint build_size_in_bits, U
 	     */
 	    Binary* bptr = erts_bin_nrml_alloc(new_size);
             ProcBin* new_pb;
+            Uint sz = PROC_BIN_SIZE;
 
             sys_memcpy(bptr->orig_bytes, binp->orig_bytes, binp->orig_size);
 
-            new_pb = (ProcBin*) HeapFragOnlyAlloc(p, PROC_BIN_SIZE);
+            /* If the subbinary is on the mature or old heap, we need to also move it */
+            if (ErtsInArea(sb, OLD_HEAP(p), OLD_HTOP(p) - OLD_HEAP(p)) ||
+                ErtsInArea(sb, HEAP_START(p), HIGH_WATER(p) - HEAP_START(p))) {
```

Comment:
ErtsInArea wants nr of **bytes**, not Eterm's.

May we need another macro ErtsInBetween(ptr, start, end).